op-node/rollup/driver/state.go

```diff
@@ -322,6 +234,34 @@ func (s *SyncDeriver) AttachEmitter(em event.Emitter) {
 
 func (s *SyncDeriver) OnEvent(ev event.Event) bool {
 	switch x := ev.(type) {
+	case status.L1UnsafeEvent:
+		// a new L1 head may mean we have the data to not get an EOF again.
+		s.Emitter.Emit(StepReqEvent{})
+	case finality.FinalizeL1Event:
+		// On "safe" L1 blocks: no step, justified L1 information does not do anything for L2 derivation or status.
+		// On "finalized" L1 blocks: we may be able to mark more L2 data as finalized now.
+		s.Emitter.Emit(StepReqEvent{})
+	case p2p.ReceivedBlockEvent:
+		envelope := x.Envelope
+		// If we are doing CL sync or done with engine syncing, fallback to the unsafe payload queue & CL P2P sync.
+		if s.SyncCfg.SyncMode == sync.CLSync || !s.Engine.IsEngineSyncing() {
+			s.Log.Info("Optimistically queueing unsafe L2 execution payload", "id", envelope.ExecutionPayload.ID())
+			s.Emitter.Emit(clsync.ReceivedUnsafePayloadEvent{Envelope: envelope})
+			s.Emitter.Emit(StepReqEvent{})
+		} else if s.SyncCfg.SyncMode == sync.ELSync {
+			ref, err := derive.PayloadToBlockRef(s.Config, envelope.ExecutionPayload)
+			if err != nil {
+				s.Log.Info("Failed to turn execution payload into a block ref", "id", envelope.ExecutionPayload.ID(), "err", err)
+				break
+			}
+			if ref.Number <= s.Engine.UnsafeL2Head().Number {
+				break
+			}
+			s.Log.Info("Optimistically inserting unsafe L2 execution payload to drive EL sync", "id", envelope.ExecutionPayload.ID())
+			if err := s.Engine.InsertUnsafePayload(s.Ctx, envelope, ref); err != nil {
+				s.Log.Warn("Failed to insert unsafe payload for EL sync", "id", envelope.ExecutionPayload.ID(), "err", err)
+			}
+		}
```

Comment:
Would be nice for readability to extract this into its own function, to keep `OnEvent` clean.

Reply:
will do